contracts/src/examples/cross-token/IpcTokenController.sol

```diff
@@ -0,0 +1,244 @@
+        address receiver,
+        uint256 amount
+    ) external payable returns (IpcEnvelope memory envelope) {
+        // Transfer and lock tokens on L1 using the inherited sendToken function
+        return _sendToken(_tokenAddr, _destinationSubnet, _destinationContract, receiver, amount);
+    }
+
+    function getUnconfirmedTransfer(bytes32 hash) public view returns (address, uint256) {
+        TransferDetails storage details = _unconfirmedTransfers[hash];
+        return (details.sender, details.value);
+    }
+
+    // Method for the contract owner to manually drop an entry from unconfirmedTransfers
+    function removeUnconfirmedTransfer(bytes32 hash) external onlyOwner {
+        _removeUnconfirmedTransfer(hash);
+    }
+
+    function _addUnconfirmedTransfer(bytes32 hash, address sender, uint256 value) internal {
+        _unconfirmedTransfers[hash] = TransferDetails(sender, value);
+    }
+
+    function _removeUnconfirmedTransfer(bytes32 hash) internal {
+        delete _unconfirmedTransfers[hash];
+    }
+
+    function _handleIpcCall(
+        IpcEnvelope memory envelope,
+        CallMsg memory callMsg
+    ) internal override validateOrigin(envelope) returns (bytes memory) {
+        bytes4 methodSignature = toBytes4(callMsg.method);
+        if (methodSignature != bytes4(keccak256("receiveAndUnlock(address,uint256)"))) {
+            revert InvalidMethod();
+        }
+
+        (address receiver, uint256 amount) = abi.decode(callMsg.params, (address, uint256));
+        // Call receiveAndUnlock to process the unlocking and transfer of tokens
+        receiveAndUnlock(receiver, amount);
+        return bytes("");
+    }
+
+    //only accept messages from replica contract
+    function verifyIpcEnvelopeLogic(IpcEnvelope memory envelope) public {
+        SubnetID memory subnetId = envelope.from.subnetId;
+        FvmAddress memory rawAddress = envelope.from.rawAddress;
+        if (!_subnetId.equals(_destinationSubnet)) {
+            revert InvalidOriginSubnet();
+        }
+        if (!rawAddress.equal(FvmAddressHelper.from(_destinationContract))) {
+            revert InvalidOriginContract();
+        }
+    }
+
+    function receiveAndUnlock(address receiver, uint256 amount) private {
+        if (receiver == address(0)) {
+            revert TransferRejected(ERR_ZERO_ADDRESS);
+        }
+
+        // Transfer the specified amount of tokens to the receiver
+        IERC20(_tokenAddr).safeTransfer(receiver, amount);
+
+        // Emit an event for the token unlock and transfer
+        emit TokensUnlocked(receiver, amount);
+    }
+
+    function _sendToken(
+        address tokenContractAddress,
+        SubnetID memory destinationSubnet,
+        address destinationContract,
+        address receiver,
+        uint256 amount
+    ) internal returns (IpcEnvelope memory committed) {
+        if (msg.value != 0) {
+            revert TransferRejected(ERR_VALUE_MUST_BE_ZERO);
+        }
```

Comment:
Is there a reason we don't validate that the token amount sent is NOT zero?

Reply:
I think we should add that check! I am going to add this as a TODO on the top and want to wait until your commits land which may have a refactor on this code